src/enzyme_ad/jax/Passes/LowerEnzymeProbProg.cpp

```diff
@@ -0,0 +1,988 @@
+
+      auto jitCall = rewriter.create<enzymexla::JITCallOp>(
+          op.getLoc(), TypeRange{loweredTraceType},
+          mlir::FlatSymbolRefAttr::get(ctx, wrapperFn), ValueRange{tracePtr},
+          rewriter.getStringAttr(""),
+          /*operand_layouts=*/
+          rewriter.getArrayAttr({rewriter.getIndexTensorAttr({})}),
+          /*result_layouts=*/
+          rewriter.getArrayAttr({rewriter.getIndexTensorAttr({})}),
+          /*output_operand_aliases=*/rewriter.getArrayAttr(aliases),
+          /*xla_side_effect_free=*/nullptr);
+
+      rewriter.replaceOp(op, jitCall.getResults());
+
+      return success();
+    } else {
+      return rewriter.notifyMatchFailure(op, "Unknown backend " + backend);
+    }
+  }
+};
+
+struct AddSampleToTraceOpConversion
+    : public OpConversionPattern<enzyme::AddSampleToTraceOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  std::string backend;
+  AddSampleToTraceOpConversion(std::string backend,
+                               TypeConverter &typeConverter,
+                               MLIRContext *context, PatternBenefit benefit = 1)
+      : OpConversionPattern(typeConverter, context, benefit), backend(backend) {
+  }
+
+  LogicalResult
+  matchAndRewrite(enzyme::AddSampleToTraceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto ctx = op->getContext();
+
+    Value trace = adaptor.getTrace();
+    auto sample = adaptor.getSample();
+
+    auto symbolWrappedAttr = op.getSymbolAttr();
+    if (!symbolWrappedAttr) {
+      return rewriter.notifyMatchFailure(op, "Missing symbol attribute");
+    }
+
+    uint64_t symbolValue = symbolWrappedAttr.getPtr();
+
+    size_t numSamples = sample.size();
+    if (numSamples == 0)
+      return rewriter.notifyMatchFailure(op, "No samples provided");
+
+    if (backend == "cpu") {
+      auto moduleOp = op->getParentOfType<ModuleOp>();
+      static int64_t fnNum = 0;
+
+      auto llvmPtrType = LLVM::LLVMPointerType::get(ctx);
+      auto llvmVoidType = LLVM::LLVMVoidType::get(ctx);
+      auto llvmI64Type = IntegerType::get(ctx, 64);
+      auto loweredTraceType = RankedTensorType::get(
+          {}, IntegerType::get(ctx, 64, IntegerType::Unsigned));
+
+      std::string addSampleToTraceFn = "enzyme_probprog_add_sample_to_trace";
+
+      auto i64TensorType = RankedTensorType::get({}, llvmI64Type);
+
+      auto symbolConst = rewriter.create<stablehlo::ConstantOp>(
+          op.getLoc(), i64TensorType,
+          cast<ElementsAttr>(makeAttr(i64TensorType, symbolValue)));
+
+      SmallVector<Type> llvmArgTypes; // (trace, symbol, n_sample_pointers...)
+      llvmArgTypes.push_back(llvmPtrType);
+      llvmArgTypes.push_back(llvmPtrType);
+      llvmArgTypes.append(numSamples, llvmPtrType);
+
+      auto funcType = LLVM::LLVMFunctionType::get(llvmVoidType, llvmArgTypes,
+                                                  /*isVarArg=*/false);
+
+      std::string wrapperFn =
```

Comment:
this can be simplified a bit, we don't need to make a different wrapper from everycallsite, we only need a different wrapper for each argument type (e.g. add f32 different from add f64)